README.md
# λ-dev

[![GitHub Actions](https://github.com/iiroj/lambda-dev/workflows/Tags/badge.svg)](https://github.com/iiroj/lambda-dev/actions)
[![version](https://img.shields.io/npm/v/lambda-dev.svg)](https://www.npmjs.com/package/lambda-dev)
[![code size](https://img.shields.io/github/languages/code-size/iiroj/lambda-dev.svg)](https://github.com/iiroj/lambda-dev)
[![dependencies](https://img.shields.io/david/iiroj/lambda-dev.svg)](https://github.com/iiroj/lambda-dev/blob/master/package.json)
[![devDependencies](https://img.shields.io/david/dev/iiroj/lambda-dev.svg)](https://github.com/iiroj/lambda-dev/blob/master/package.json)

Develop and Bundle Your Lambda Functions With Ease.

## Installation

`npm install --save-dev lambda-dev`

`yarn add --dev lambda-dev`

## Usage

### Development

Use `lambda-dev` to develop lambda functions locally. `lambda-dev serve` starts an [Express](https://expressjs.com) server that proxies http requests to your lambda functions. They are transpiled with [@babel/core](https://babeljs.io/docs/en/next/babel-core) and [@babel/preset-env](https://babeljs.io/docs/en/next/babel-preset-env), with the node target set to `--node [target]` (default `6.10`). This is done through [webpack](https://webpack.js.org) with the help of [babel-loader](https://github.com/babel/babel-loader).

```bash
lambda-dev serve --help
lambda-dev serve src/functions --node 8.10 --port 9000 --basePath /lambda
```

Now a given function `src/functions/test.js` will be invoked with requests to `http://localhost:9000/lambda/test`.

### Build

```bash
lambda-dev build --help
lambda-dev serve src/functions build/functions --node 8.10
```

Bundled functions will be at `build/functions`.

## Custom Webpack configuration

It is possible to supply a custom Webpack configuration for serving and building your Lambda functions:

```bash
lambda-dev serve src/functions --webpack-config ./my-webpack.config.js
```

where the default export of `my-webpack.config.js` should be either and object or a function. If it's an object, it will be merged with the default configuration using `merge.smart` from [webpack-merge](https://github.com/survivejs/webpack-merge). If it's a function, it will receive the default configuration as its argument and should return a full valid configuration.

## Custom babel configuration

It's not possible to directly supply a custom babel configuration, but you can override the webpack configuration's [babel-loader](https://github.com/babel/babel-loader) options:

```javascript
const myBabelOptions = require('./my-babel.config.js');

module.exports = {
  module: {
    rules: [
      {
        test: /\.js$/,
        use: {
          loader: require.resolve('babel-loader'),
          options: {
            babelrc: false,
            options: myBabelOptions
          }
        }
      }
    ]
  }
};
```

## Lambda Function Specification

### [Read the official docs](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-handler.html)

Lambda functions should export a `handler` function that receives the following arguments:

```javascript
import { Request } from 'express';

type Event = {
  path: Request.path,
  httpMethod: Request.method,
  queryStringParameters: Request.query,
  headers: Request.headers,
  body: Request.body
};

type Context = {} // Empty with `lambda-dev serve`

exports.handler: (event: Event, context: Context, callback) => callback(error: Error | null, response: Response | null);
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.8.22](https://github.com/iiroj/lambda-dev/compare/v1.8.21...v1.8.22) (2019-12-30)

### [1.8.21](https://github.com/iiroj/lambda-dev/compare/v1.8.20...v1.8.21) (2019-10-26)

### [1.8.20](https://github.com/iiroj/lambda-dev/compare/v1.8.19...v1.8.20) (2019-09-25)

### [1.8.19](https://github.com/iiroj/lambda-dev/compare/v1.8.18...v1.8.19) (2019-09-22)

### [1.8.18](https://github.com/iiroj/lambda-dev/compare/v1.8.17...v1.8.18) (2019-08-27)

### [1.8.17](https://github.com/iiroj/lambda-dev/compare/v1.8.16...v1.8.17) (2019-07-29)



### [1.8.16](https://github.com/iiroj/lambda-dev/compare/v1.8.15...v1.8.16) (2019-07-11)



### [1.8.15](https://github.com/iiroj/lambda-dev/compare/v1.8.14...v1.8.15) (2019-07-07)


### Build System

* add husky, commitlint and lint-staged ([ea89cd7](https://github.com/iiroj/lambda-dev/commit/ea89cd7))



### [1.8.14](https://github.com/iiroj/lambda-dev/compare/v1.8.13...v1.8.14) (2019-07-06)



### [1.8.13](https://github.com/iiroj/lambda-dev/compare/v1.8.12...v1.8.13) (2019-06-02)


### Bug Fixes

* update types ([8225076](https://github.com/iiroj/lambda-dev/commit/8225076))



## [1.8.12](https://github.com/iiroj/lambda-dev/compare/v1.8.11...v1.8.12) (2019-04-28)



## [1.8.11](https://github.com/iiroj/lambda-dev/compare/v1.8.10...v1.8.11) (2019-03-27)



## [1.8.10](https://github.com/iiroj/lambda-dev/compare/v1.8.9...v1.8.10) (2019-03-17)



## [1.8.9](https://github.com/iiroj/lambda-dev/compare/v1.8.8...v1.8.9) (2019-03-02)



## [1.8.8](https://github.com/iiroj/lambda-dev/compare/v1.8.7...v1.8.8) (2019-03-02)



## [1.8.7](https://github.com/iiroj/lambda-dev/compare/v1.8.6...v1.8.7) (2019-03-02)



## [1.8.6](https://github.com/iiroj/lambda-dev/compare/v1.8.5...v1.8.6) (2019-03-02)


### Bug Fixes

* **webpack:** better fix for empty entry names falling back to “index” ([ac3eb54](https://github.com/iiroj/lambda-dev/commit/ac3eb54))



## [1.8.5](https://github.com/iiroj/lambda-dev/compare/v1.8.4...v1.8.5) (2019-03-02)


### Bug Fixes

* **webpack:** empty entry names fallback to “index” ([26bd1f6](https://github.com/iiroj/lambda-dev/commit/26bd1f6))



## [1.8.4](https://github.com/iiroj/lambda-dev/compare/v1.8.3...v1.8.4) (2019-03-02)



## [1.8.3](https://github.com/iiroj/lambda-dev/compare/v1.8.2...v1.8.3) (2019-03-02)



## [1.8.2](https://github.com/iiroj/lambda-dev/compare/v1.8.1...v1.8.2) (2019-02-23)



<a name="1.8.1"></a>
## [1.8.1](https://github.com/iiroj/lambda-dev/compare/v1.8.0...v1.8.1) (2019-02-05)



<a name="1.8.0"></a>
# [1.8.0](https://github.com/iiroj/lambda-dev/compare/v1.8.0-0...v1.8.0) (2019-01-30)



<a name="1.8.0-0"></a>
# [1.8.0-0](https://github.com/iiroj/lambda-dev/compare/v1.7.2...v1.8.0-0) (2019-01-29)


### Features

* purge node’s require cache of changed files ([9f51e7e](https://github.com/iiroj/lambda-dev/commit/9f51e7e))



<a name="1.7.2"></a>
## [1.7.2](https://github.com/iiroj/lambda-dev/compare/v1.7.1...v1.7.2) (2019-01-26)



<a name="1.7.1"></a>
## [1.7.1](https://github.com/iiroj/lambda-dev/compare/v1.7.0...v1.7.1) (2019-01-26)



<a name="1.7.0"></a>
# [1.7.0](https://github.com/iiroj/lambda-dev/compare/v1.6.0...v1.7.0) (2019-01-26)


### Features

* rewrite in TypeScript and bundle using Rollup ([9cfd2e1](https://github.com/iiroj/lambda-dev/commit/9cfd2e1))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/iiroj/lambda-dev/compare/v1.5.10...v1.6.0) (2019-01-23)


### Features

* Support isBase64Encoded ([b51f0ce](https://github.com/iiroj/lambda-dev/commit/b51f0ce))



<a name="1.5.10"></a>
## [1.5.10](https://github.com/iiroj/lambda-dev/compare/v1.5.9...v1.5.10) (2019-01-19)


### Bug Fixes

* serve correct handler for route by requestPath ([f8b5eb0](https://github.com/iiroj/lambda-dev/commit/f8b5eb0))



<a name="1.5.9"></a>
## [1.5.9](https://github.com/iiroj/lambda-dev/compare/v1.5.8...v1.5.9) (2019-01-14)


### Bug Fixes

* log correct warning ([639ca89](https://github.com/iiroj/lambda-dev/commit/639ca89))



<a name="1.5.8"></a>
## [1.5.8](https://github.com/iiroj/lambda-dev/compare/v1.5.7...v1.5.8) (2019-01-13)



<a name="1.5.7"></a>
## [1.5.7](https://github.com/iiroj/lambda-dev/compare/v1.5.6...v1.5.7) (2019-01-13)



<a name="1.5.6"></a>
## [1.5.6](https://github.com/iiroj/lambda-dev/compare/v1.5.5...v1.5.6) (2019-01-13)



<a name="1.5.5"></a>
## [1.5.5](https://github.com/iiroj/lambda-dev/compare/v1.5.4...v1.5.5) (2019-01-13)



<a name="1.5.4"></a>
## [1.5.4](https://github.com/iiroj/lambda-dev/compare/v1.5.3...v1.5.4) (2019-01-09)



<a name="1.5.3"></a>
## [1.5.3](https://github.com/iiroj/lambda-dev/compare/v1.5.2...v1.5.3) (2019-01-09)



<a name="1.5.2"></a>
## [1.5.2](https://gitlab.com/iiroj/lambda-dev/compare/v1.5.1...v1.5.2) (2019-01-03)



<a name="1.5.1"></a>
## [1.5.1](https://gitlab.com/iiroj/lambda-dev/compare/v1.5.0...v1.5.1) (2018-12-21)


### Bug Fixes

* handle cache busting for index routes ([4d1f216](https://gitlab.com/iiroj/lambda-dev/commit/4d1f216))
* log served function names only once ([dc2b424](https://gitlab.com/iiroj/lambda-dev/commit/dc2b424))



<a name="1.5.0"></a>
# [1.5.0](https://gitlab.com/iiroj/lambda-dev/compare/v1.4.10...v1.5.0) (2018-12-21)


### Features

* add --include and --exclude options for globbing functions ([fc710d2](https://gitlab.com/iiroj/lambda-dev/commit/fc710d2))



<a name="1.4.10"></a>
## [1.4.10](https://gitlab.com/iiroj/lambda-dev/compare/v1.4.9...v1.4.10) (2018-11-30)



<a name="1.4.9"></a>
## [1.4.9](https://gitlab.com/iiroj/lambda-dev/compare/v1.4.8...v1.4.9) (2018-11-10)



<a name="1.4.8"></a>
## [1.4.8](https://gitlab.com/iiroj/lambda-dev/compare/v1.4.7...v1.4.8) (2018-10-29)



<a name="1.4.7"></a>
## [1.4.7](https://gitlab.com/iiroj/lambda-dev/compare/v1.4.6...v1.4.7) (2018-10-12)



<a name="1.4.6"></a>

src/cli.ts
import * as yargs from "yargs";

import serveHandler from "./serve";
import buildHandler from "./build";

yargs
  .demandCommand(1)
  .option("include", {
    alias: "i",
    describe: "Additional glob for including Lambda Functions",
    type: "array"
  })
  .option("exclude", {
    alias: "e",
    describe: "Glob for excluding Lambda Functions from the included set",
    type: "array"
  })
  .option("node", {
    alias: "n",
    default: "8.10",
    description: "Target Node.js version, used with @babel/preset-env",
    type: "string"
  })
  .option("webpack-config", {
    alias: "w",
    description:
      "Relative path to optional Webpack configuration. The default export should be an object that is merged with the default webpack config, or a function that receives the default configuration and returns a valid webpack configuration.",
    type: "string"
  })
  .option("watch", {
    description:
      "Control the webpack watch mode. Defaults to true for serve and false for build.",
    type: "boolean"
  })
  .command<CliServeArgs>({
    command: "serve <entry>",
    describe: "Lambda Development Server",
    builder: (yargs: any) =>
      yargs
        .positional("entry", {
          describe: "Entry glob for Lambda functions"
        })
        .option("basePath", {
          alias: "b",
          default: "/",
          description: "Base pathname prepended to function filenames",
          type: "string"
        })
        .option("port", {
          alias: "p",
          default: 9000,
          description: "Port for development server",
          type: "number"
        }),
    handler: serveHandler
  })
  .command<CliBuildArgs>({
    command: "build <entry> <target>",
    describe: "Bundle Lambda functions",
    builder: (yargs: any) =>
      yargs
        .positional("entry", {
          describe: "Relative path to entry file or directory of functions",
          type: "string"
        })
        .positional("target", {
          describe: "Relative path to target directory for bundled functions",
          type: "string"
        }),
    handler: buildHandler
  }).argv;
